### PR TITLE
test_shut_down.py: assert reboot flushes metrics

### PR DESCRIPTION
**Pull request by iulianbarbu**
## Reason for This PR

#1624

## Description of Changes

Updated `test_shut_down.py` to assert the flushing of metrics when a reboot is issued from within the booted microVM.

`mod tests` from vmm/lib.rs is not neccessary for this file, since the units testing fit better for integration tests, which are already covered by our current integration tests.

Modified the API for creation and flags manipulation of a `logging::Fifo`. These changes offer flexibility in choosing if the opened pipe is blocking or non blocking and support for changing file descriptor flags for the opened fifo.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria. Where there are two options, keep one.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] This PR is linked to an issue.
- [x] The description of changes is clear and encompassing.
- [x] Bo docs need to be updated as part of this PR.
- [x] Code-level documentation for touched code is included in this PR.
- [x] No API changes are included in this PR.
- [x] The changes in this PR have no user impact.
- [x] No new `unsafe` code has been added.

### CODE REVIEW

**Comment by dianpopa** [comment]
I do think there are some potential scenarios that vmm/src/lib.rs should get unit tested/ci tested against:

==Unit tests==
* `get_bus_device( device_type: DeviceType,
        device_id: &str,)` -> isn 't it worth it to test that not registering a (device_type, id) tuple would make this function return `None`? Or also what about making sure that registering a net device with the same id as a block device works correctly?
* `guest_memory` -> why not test that it is actually returns what we configured?
*  `pub fn kvm_vm(&self)` is not used at all

==CI==
* Are  metrics flushed upon  `pub fn stop(&mut self, exit_code: i32)` ?
* Does `configure_system` fail if there is not enough memory to setup the mptable?

**Comment by iulianbarbu** [comment]
> I do think there are some potential scenarios that vmm/src/lib.rs should get unit tested/ci tested against:
> 
> ==Unit tests==
> 
>     * `get_bus_device( device_type: DeviceType, device_id: &str,)` -> isn 't it worth it to test that not registering a (device_type, id) tuple would make this function return `None`? Or also what about making sure that registering a net device with the same id as a block device works correctly?

Testing this unit means testing the `MmioDeviceManager::get_bus_device` (because it is a wrapper which does not have any related logic), which is covered by `device_manager/mmio.rs::test_device_info` unit test. I genuinely do no think it would bring any value to have a unit test for this function.

> 
>     * `guest_memory` -> why not test that it is actually returns what we configured?
> 
>     * `pub fn kvm_vm(&self)` is not used at all

Unit testing for getters, if they do not have any complex logic, it would require enough effort just for nothing, in my opinion. I've excluded them from the report (#1624) just because they lack any related logic that would bring value for a unit test.
 
> 
> ==CI==
> 
>     * Are  metrics flushed upon  `pub fn stop(&mut self, exit_code: i32)` ?

I think this is a scenario that can be tested inside an integration testing. I will add an integration test for this. 

>     * Does `configure_system` fail if there is not enough memory to setup the mptable?

`vmm/lib.rs::configure_system` is a wrapper over `arch_*::configure_system`, without any related logic to it, besides effective calls to `arch_*::configure_system`. Testing it for `NotEnoughMemory` for mptable means testing `configure_system` from `arch_*`. This is already covered in unit tests.

**Comment by dianpopa** [comment]
> > I do think there are some potential scenarios that vmm/src/lib.rs should get unit tested/ci tested against:
> > ==Unit tests==
> > ```
> > * `get_bus_device( device_type: DeviceType, device_id: &str,)` -> isn 't it worth it to test that not registering a (device_type, id) tuple would make this function return `None`? Or also what about making sure that registering a net device with the same id as a block device works correctly?
> > ```
> 
> Testing this unit means testing the `MmioDeviceManager::get_bus_device` (because it is a wrapper which does not have any related logic), which is covered by `device_manager/mmio.rs::test_device_info` unit test. I genuinely do no think it would bring any value to have a unit test for this function.
> 
> > ```
> > * `guest_memory` -> why not test that it is actually returns what we configured?
> > 
> > * `pub fn kvm_vm(&self)` is not used at all
> > ```
> 
> Unit testing for getters, if they do not have any complex logic, it would require enough effort just for nothing, in my opinion. I've excluded them from the report (#1624) just because they lack any related logic that would bring value for a unit test.
> 
> > ==CI==
> > ```
> > * Are  metrics flushed upon  `pub fn stop(&mut self, exit_code: i32)` ?
> > ```
> 
> I think this is a scenario that can be tested inside an integration test. I will add an integration test for this.
> 
> > ```
> > * Does `configure_system` fail if there is not enough memory to setup the mptable?
> > ```
> 
> `vmm/lib.rs::configure_system` is a wrapper over `arch_*::configure_system`, without any related logic to it, besides effective calls to `arch_*::configure_system`. Testing it for `NotEnoughMemory` for mptable means testing `configure_system` from `arch::x86_64`. This is already covered in unit tests.

**Comment by andreeaflorescu** [comment]
> I am not on the same page here regarding not making sense to test simple bits of code. Unit tests are not made only for complex units. Testing a getter ensures that the getter stays a getter. Unit tests should also be seen as an extra layer of protection against malicious or unintentional mistakes in PRs.

I agree with you that you generally want to test even the simple bits of code. In this particular usecase, testing the `Vmm` requires a lot of setup and comes with tests that are hard to maintain. The value added by testing setters and getters tends to fade away when taking into consideration the long time required for maintaining the tests and their setup. I think it makes more sense to invest this time in writing integration tests. WDYT?

**Comment by dianpopa** [comment]
> > I am not on the same page here regarding not making sense to test simple bits of code. Unit tests are not made only for complex units. Testing a getter ensures that the getter stays a getter. Unit tests should also be seen as an extra layer of protection against malicious or unintentional mistakes in PRs.
> 
> I agree with you that you generally want to test even the simple bits of code. In this particular usecase, testing the `Vmm` requires a lot of setup and comes with tests that are hard to maintain. The value added by testing setters and getters tends to fade away when taking into consideration the long time required for maintaining the tests and their setup. I think it makes more sense to invest this time in writing integration tests. WDYT?

I think we should invest time in writing integration tests regardless of whether we write unit tests for the vmm or not.
As far as my concern, no exception should be made as far as unit tests. It is a dangerous direction to start being subjective about what you unit test and what you do not. Unit test should exercise exactly a single method from a public API regardless of what it is built upon or what calls it. That is why declaring the `impl vmm` as unit tested is wrong IMO.

However, this PR can go through.

File: tests/integration_tests/functional/test_metrics.py [hunk 1]
```diff
@@ -33,3 +35,55 @@ def test_flush_metrics(test_microvm_with_api):
...
+    metrics_file_path = os.path.join(microvm.path, 'metrics_file')
+    metrics_file = log_tools.File(metrics_file_path)
+
+    response = microvm.metrics.put(
+        metrics_fifo=microvm.create_jailed_resource(metrics_file.path)
+    )
+    assert microvm.api_session.is_status_no_content(response.status_code)
+
+    microvm.start()
+    time.sleep(2)
```
Comment (acatangiu) [inline tests/integration_tests/functional/test_metrics.py]: I don't think this is robust code, sleeping for two seconds is arbitrarily too long or too short depending on the machine this test runs and/or how loaded the machine is.

It would be better to use a solution that deterministic-ally finds when a microVM has booted, then continues running the test.

**Comment by aghecenco** [comment]
Re: the unit testing of small bits of code: I agree with @dianpopa that we should test everything, even if it's already covered. I've seen way too many getters that were sneaking in a bit of logic on the side before returning a value. I believe that code that isn't thoroughly tested has a better chance of not aging well.

File: tests/integration_tests/functional/test_metrics.py [hunk 2]
```diff
@@ -3,7 +3,9 @@
 """Tests the metrics system."""
 
 import os
-
+import time
```
Comment (aghecenco) [inline tests/integration_tests/functional/test_metrics.py]: nit: please group imports alphabetically.

File: tests/integration_tests/functional/test_metrics.py [hunk 3]
```diff
@@ -33,3 +35,55 @@ def test_flush_metrics(test_microvm_with_api):
...
     lines = metrics_fifo.sequential_reader(how_many_flushes)
     assert len(lines) == how_many_flushes
+
+
+@pytest.mark.skipif(
+    platform.machine() != "x86_64",
+    reason="not yet implemented on aarch64"
+)
+def test_stop_and_flush_metrics(test_microvm_with_atkbd):
+    """Check if stop/reboot flush the metrics."""
```
Comment (aghecenco) [inline tests/integration_tests/functional/test_metrics.py]:
```suggestion
    """Check if stop/reboot flushes the metrics."""
```

File: tests/host_tools/logging.py [hunk 4]
```diff
@@ -11,32 +11,32 @@
 from threading import Thread
 
 
-class Fifo:
-    """Facility for creating and working with named pipes (FIFOs)."""
+class File:
```
Comment (aghecenco) [inline tests/host_tools/logging.py]: Why this change? Why not keep using fifos? I think we should stick to fifos because in the long run we'll add high-density tests. Fifos at least have limited capacity, so, if nobody's reading from the other end, they won't accept new incoming data. Files on the other hand will fill up the disk if anything's wrong. If there's a synchronization or I/O issue with fifos, please fix that instead.

Comment (acatangiu) [reply tests/host_tools/logging.py]: Disclaimer: I'm fine with both files or fifos here.

There are pros and cons to each, but just to make sure we care about the right ones, will this code ever be used for high-density testing (honest question)?

**Comment by iulianbarbu** [comment]
@dianpopa , @acatangiu , @aghecenco , I think I've managed your concerns. Please take a look again.

File: tests/host_tools/logging.py [hunk 5]
```diff
@@ -2,35 +2,43 @@
...
-        fifo = self._open_nonblocking()
-        return fifo.readlines()[:max_lines]
+        return self.fifo.readlines()[:max_lines]
+
+    def sequential_blocking_reader(self, max_lines):
+        """Return up to `max_lines` lines from fifo which blocks.
+
+        :return: A list containing the read lines.
+        """
+        blocking_fifo = open(self.path, "r")
```
Comment (aghecenco) [inline tests/host_tools/logging.py]: Why open a new one instead of `fcntl`'ing the existing fd?

Comment (aghecenco) [reply tests/host_tools/logging.py]: Actually, why not specify in the constructor if you want the fifo to be blocking or not, and use a `self.blocking` attribute to control the behavior?

File: tests/host_tools/logging.py (hunk 4 above)
Comment (aghecenco) [reply tests/host_tools/logging.py]: I'd see it that way, yup. I could see the use of throughput metrics in high density, as well as the boot time log (which we looked for in the `test_multiple_microvms` scenarios). So unless we rewrite the whole test harness (or write a new one for perf&stress), I think this framework might be used in density tests too.

File: tests/host_tools/logging.py (hunk 5 above)
Comment (iulianbarbu) [reply tests/host_tools/logging.py]: Yes, specifying the blocking behavior in the constructor makes sense. I will add also a method for changing the blocking behavior. E.g. opening a pipe in blocking mode, use it, then deciding that we want to use it later on with non blocking behavior, or vice versa. WDYT?

File: tests/host_tools/logging.py [hunk 6]
```diff
@@ -15,23 +16,45 @@ class Fifo:
...
+        """
+        fd = self.fifo.fileno()
+        return fcntl.fcntl(fd, fcntl.F_GETFL)
+
+    def or_flags(self, flags):
+        """Or between `flags` arg and opened filed existing flags."""
+        new_flags = self.flags() | flags
+        self.set_flags(new_flags)
+
+    def set_flags(self, flags):
```
Comment (aghecenco) [inline tests/host_tools/logging.py]: The getter/setter could be replaced with a `@property` like here: https://github.com/firecracker-microvm/firecracker/blob/eef40ed38718b8ef065a7bf2d53c64b30ffafb62/tests/framework/microvm.py#L169

Comment (iulianbarbu) [reply tests/host_tools/logging.py]: @aghecenco , please take a look! I've added @property decorator, as suggested.

Comment (dianpopa) [reply tests/host_tools/logging.py]: The nice thing about this type of decorators is that you can simply do:
```
# getter example
fl  = fifo.flags 
# setter
fifo.flags = OS_NONBLOCK
```
So the perk here is that you can use same name for both getter and setter when decorating them. However, I see you used "set_flags" for the setter.

Comment (iulianbarbu) [reply tests/host_tools/logging.py]: Oh, right. I will push pushed an updated with your usage. Thanks.
